### PR TITLE
ci: add workflows

### PR DESCRIPTION
**Pull request by JaeAeich**
## Summary by Sourcery

Introduce a new Task Execution Service (TES) module along with supporting modules for transport, configuration, and service information. Add comprehensive CI workflows for PR evaluation, code testing, and code quality checks. Implement build and test scripts, and add documentation and issue templates.

New Features:
- Introduce a new module for Task Execution Service (TES) with functionalities to create, get, list, and cancel tasks.
- Add a new Transport module to handle HTTP requests with support for GET, POST, PUT, and DELETE methods.
- Implement a new Configuration module to manage API configurations including base path, user agent, and authentication details.
- Add a ServiceInfo module to fetch service information using the new Transport module.

Enhancements:
- Add a build script (build-models.sh) to generate models from OpenAPI specifications.
- Introduce a pre-commit configuration (.pre-commit-config.yaml) to enforce code quality checks.
- Add a markdown lint configuration (.markdownlint.yaml) to enforce consistent markdown styling.
- Add a YAML lint configuration (.yamllint.yaml) to enforce consistent YAML styling.
- Add a nextest configuration (nextest.toml) to manage test retries and setup commands.

Build:
- Add a build script (build-models.sh) to automate the generation of models from OpenAPI specifications.

CI:
- Add a CI workflow (.github/workflows/pr_evaluation.yaml) to evaluate pull requests with checks for merge conflicts, pre-commit hooks, semantic PR titles, and binary builds.
- Add a CI workflow (.github/workflows/code_test.yaml) to run integration and unit tests with coverage reporting.
- Add a CI workflow (.github/workflows/code_quality.yaml) to run formatting, linting, and spell check tasks.

Documentation:
- Add a README.md file with instructions for building the project, running tests, and testing the CI/CD workflow locally.

Tests:
- Add unit tests for the TES module to test task creation, status retrieval, cancellation, and listing.
- Add unit tests for the Transport module to test HTTP request handling.
- Add a test utility module (test_utils.rs) to setup the testing environment and ensure Funnel server is running.

Chores:
- Add a general-purpose issue template (.github/ISSUE_TEMPLATE/general-purpose.md) for reporting issues.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/general-purpose.md
Comment (uniqueg) [inline .github/ISSUE_TEMPLATE/general-purpose.md:1]: Doesn't have anything to do with CI?

File: .github/workflows/code_quality.yaml
Comment (uniqueg) [inline .github/workflows/code_quality.yaml:43]: Run linting?

File: .github/workflows/code_test.yaml
Comment (uniqueg) [inline .github/workflows/code_test.yaml:27]: Add to `Cargo.toml`?

File: .github/workflows/pr_evaluation.yaml
Comment (uniqueg) [inline .github/workflows/pr_evaluation.yaml:36]: > Check that all pre-commit hooks passed

**Review by uniqueg — CHANGES_REQUESTED** [review]
Some minor comments, but generally the workflows look really good/useful. Of course needs some changes in the code to pass. Also, perhaps a more Rust-experienced person could look at it, because I don't know the toolchain really. @pavelnikonorov @vemonet 

I have added `CODECOV_TOKEN`.

File: .github/ISSUE_TEMPLATE/general-purpose.md
Comment (JaeAeich) [reply .github/ISSUE_TEMPLATE/general-purpose.md:1]: :sob: yeah I just wanted to add most of the things that were left from cookiecutter thats it, not just the CI.

File: .github/workflows/code_quality.yaml
Comment (JaeAeich) [reply .github/workflows/code_quality.yaml:43]: we have format just above that don't we, what do you mean?

File: .github/workflows/code_test.yaml
Comment (JaeAeich) [reply .github/workflows/code_test.yaml:27]: aah yes, didn't touch that as it wan't merged.

**Comment by JaeAeich** [comment]
@uniqueg either merge this after we have merged all other PRs, because resolving convo will mean me changing some code so it doesn't make sense to do that rn, or merge this into some PR take it from there which might not be ideal though.

File: .github/workflows/pr_evaluation.yaml
Comment (JaeAeich) [reply .github/workflows/pr_evaluation.yaml:36]: :exploding_head: this is what I have in cookiecutter though!!!! Do I have change it there as well?

File: .github/workflows/code_test.yaml
Comment (vemonet) [reply .github/workflows/code_test.yaml:27]: I think since it is a CLI tool it needs to be installed manually outside of the `Cargo.toml`, or maybe you can add it to the rust-toolchain file

File: .github/workflows/pr_evaluation.yaml
Comment (vemonet) [reply .github/workflows/pr_evaluation.yaml:36]: Why the need for poetry here?

File: .github/workflows/code_quality.yaml
Comment (vemonet) [inline .github/workflows/code_quality.yaml:28]: The 2 first lines seems to be running the actual formatting, are they really needed? It seems like we should only be doing fmt check in the workflow no?

File: .github/workflows/code_test.yaml
Comment (vemonet) [inline .github/workflows/code_test.yaml:30]: I think you should also add the `--doc` flag so that tarpaulin includes codeblocks from the docstrings

**Comment by vemonet** [comment]
Looking good, most things required seems to be there: tests, cov, fmt, linting, precommit

**Comment by JaeAeich** [comment]
> Looking good, most things required seems to be there: tests, cov, fmt, linting, precommit

Thanks for the review, Ill make changes when the branch on which this is bases off of gets merged.

File: .github/ISSUE_TEMPLATE/general-purpose.md
Comment (uniqueg) [reply .github/ISSUE_TEMPLATE/general-purpose.md:1]: Fair. But let's make sure to remove this from this PR and add it to another (together with the PR template).

File: .github/workflows/code_test.yaml
Comment (uniqueg) [reply .github/workflows/code_test.yaml:27]: Really? What's the difference? Aren't the deps in `Cargo.toml` installed via `cargo install`?

File: .github/workflows/pr_evaluation.yaml
Comment (JaeAeich) [reply .github/workflows/pr_evaluation.yaml:36]: Please ignore, was copying the structure from a python project, must have been left in. Will remove.

Comment (uniqueg) [reply .github/workflows/pr_evaluation.yaml:36]: Yes, not a biggie of course. You can change it here, and sneak it in some other time in the cookiecutter template, if you remember

Comment (uniqueg) [reply .github/workflows/pr_evaluation.yaml:36]: Ah, yes, overlooked the Poetry artifacts here - good catch @vemonet

**Comment by uniqueg** [comment]
> > Looking good, most things required seems to be there: tests, cov, fmt, linting, precommit
> 
> Thanks for the review, Ill make changes when the branch on which this is bases off of gets merged.

"This branch has no conflicts with the base branch" <- I think you could already make the changes? Of course, it may need some other stuff until tests pass, but I do think you could work towards resolving the open issues at this point, @JaeAeich and @aaravm

**Review by pavelnikonorov — APPROVED** [review]
@uniqueg I've reviewed all conversations, and committed the discussed changes. 
also, I have tested it locally using `act`, the workflows work, but the tests fail, which should be addressed separately and within the other PRs review.

Added two issues out of this PR: https://github.com/elixir-cloud-aai/ga4gh-sdk/issues/44, https://github.com/elixir-cloud-aai/ga4gh-sdk/issues/45

P.S. FYI, I've done a few push-force on my commits here to amend them, but it appears that it uses a different credentials source, which belongs to Grigorii who used to work on the same machine.

**Comment by pavelnikonorov** [comment]
all suggestions addressed, merging as discussed with @uniqueg

**Pull request merged**
